parlai/agents/hugging_face/llama.py

```diff
@@ -0,0 +1,380 @@
+
+    def _tie_weights(self, lm_head, output_embeddings):
+        lm_head.weight = output_embeddings.weight
+
+    def reorder_encoder_states(self, encoder_states, indices):
+        enc = torch.index_select(encoder_states, 0, indices)
+        return enc
+
+    def output(self, tensor):
+        """
+        Compute output logits.
+        """
+        return self.lm_head(tensor)
+
+    def reorder_decoder_incremental_state(self, incremental_state, inds):
+        new_incr_state = []
+        for layer_past in incremental_state:
+            if torch.is_tensor(layer_past):
+                new_incr_state.append(torch.index_select(layer_past, 1, inds))
+            else:
+                # newer versions of HF split up the intermediate outputs
+                assert isinstance(layer_past, tuple)
+                layer_past = torch.stack(layer_past, dim=0)
+                new_incr_state.append(torch.index_select(layer_past, 1, inds))
+
+        return tuple(new_incr_state)
+
+
+############################################
+## Generative Agent
+############################################
+
+
+class LlamaAgent(TorchGeneratorAgent):
+    """
+    Hugging Face Llama Agent.
+
+    LlamaForCausalLM is a multi-layer decoder-only Transformer.
+    The decoder is initialized with pretrained weights from Hugging Face.
+    Read more about this model here
+    <https://huggingface.co/transformers/model_doc/llama.html>.
+
+    Llama comes in multiple sizes: 1.5B, 2.7B, 7B 30B, 65B. Use the
+    flag `--llama-size` to choose the size.
+    """
+
+    @classmethod
+    def add_cmdline_args(
+        cls, parser: ParlaiParser, partial_opt: Optional[Opt] = None
+    ) -> ParlaiParser:
+        super().add_cmdline_args(parser, partial_opt=partial_opt)
+        group = parser.add_argument_group("LlamaAgent Args")
+        group = setup_llama_args(group)
+        return parser
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        if hasattr(self.model, "module"):
+            self.START_IDX = self.model.module.START_IDX
+            self.END_IDX = self.model.module.END_IDX
+            self.NULL_IDX = self.model.module.NULL_IDX
+        else:
+            self.START_IDX = self.model.START_IDX
+            self.END_IDX = self.model.END_IDX
+            self.NULL_IDX = self.model.NULL_IDX
+
+    @staticmethod
+    def dictionary_class():
+        """
+        Return the dictionary class that this agent expects to use.
+
+        Can be overridden if a more complex dictionary is required.
+        """
+        return LlamaDictionaryAgent
+
+    def build_model(self, states=None):
+        """
+        Build and return model.
+        """
+        return ParlaiLlamaForCausalLM(self.opt, self.dict)
+
+    def _encoder_input(self, batch):
+        return (batch.text_vec,)
+
+    def _pad_tensor(self, items, is_label=False):
+        """
+        Override to always set fp16friendly to False.
```

Comment:
why?

Reply:
should I (set fp16friendly to False?) saw that from gpt2 agent @klshuster